### PR TITLE
Add MOVE_SHELL_TRAP Effect

### LINKED ISSUE #2476: MOVE_SHELL_TRAP Effect

### PR DESCRIPTION
**Pull request by DizzyEggg**
Fixes #2476

Replaces #2132

I used AgustinGDLV's branch as a base and worked on that. Everything should be covered here.
Also added tests for Shell Trap effect.

### CODE REVIEW

File: src/battle_script_commands.c [hunk 1]
```diff
@@ -5725,6 +5726,22 @@ static void Cmd_moveend(void)
             else
                 gBattleStruct->lastMoveFailed &= ~(gBitTable[gBattlerAttacker]);
 
+            // Set ShellTrap to activate after the attacker's turn if target was hit by a physical move.
+            if (gBattleMoves[gBattleMons[gBattlerTarget].moves[gBattleStruct->chosenMovePositions[gBattlerTarget]]].effect == EFFECT_SHELL_TRAP
```
Comment (mrgriffin) [inline src/battle_script_commands.c:5730]: This is ridiculously narrow, but in a double-battle where I KO one of the foes and replace the move I chose with Shell Trap, could it then activate on the same turn despite not having been chosen?

**Review by mrgriffin — COMMENTED** [review]
Love the tests :)

File: src/battle_script_commands.c (hunk 1 above)
Comment (DizzyEggg) [reply src/battle_script_commands.c:5730]: You're totally right, I think this is exactly what would happen. I think it could be easily fixed tho

Comment (DizzyEggg) [reply src/battle_script_commands.c]: Fixed

**Review by mrgriffin — APPROVED** [review]

File: test/move_effect_shell_trap.c [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+SINGLE_BATTLE_TEST("Shell Trap activates only if hit by a physical move")
+{
+    u32 move;
+    bool32 activate;
+    PARAMETRIZE { move = MOVE_TACKLE; activate = TRUE; }
+    PARAMETRIZE { move = MOVE_WATER_GUN; activate = FALSE; }
+    PARAMETRIZE { move = MOVE_LEER; activate = FALSE; }
+
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET) {}
```
Comment (mrgriffin) [inline test/move_effect_shell_trap.c:21]:
```suggestion
        PLAYER(SPECIES_WOBBUFFET);
```

File: test/move_effect_shell_trap.c [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    u32 move;
+    bool32 activate;
+    PARAMETRIZE { move = MOVE_TACKLE; activate = TRUE; }
+    PARAMETRIZE { move = MOVE_STOMP; activate = FALSE; }
+
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_TAUROS) { Ability(ABILITY_SHEER_FORCE); }
+    } WHEN {
+        TURN {MOVE(player, MOVE_SHELL_TRAP);  MOVE(opponent, move); }
```
Comment (mrgriffin) [inline test/move_effect_shell_trap.c:56]:
```suggestion
        TURN { MOVE(player, MOVE_SHELL_TRAP); MOVE(opponent, move); }
```

File: test/move_effect_shell_trap.c [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                HP_BAR(opponent);
+            }
+        }
+    }
+}
+
+SINGLE_BATTLE_TEST("Shell Trap does not activate if battler faints before being able to activate it")
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET) {HP(1); }
```
Comment (mrgriffin) [inline test/move_effect_shell_trap.c:79]:
```suggestion
        PLAYER(SPECIES_WOBBUFFET) { HP(1); }
```

File: test/move_effect_shell_trap.c [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN {MOVE(player, MOVE_SHELL_TRAP);  MOVE(opponent, MOVE_TACKLE); SEND_OUT(player, 1); }
```
Comment (mrgriffin) [inline test/move_effect_shell_trap.c:83]:
```suggestion
        TURN { MOVE(player, MOVE_SHELL_TRAP); MOVE(opponent, MOVE_TACKLE); SEND_OUT(player, 1); }
```

File: test/move_effect_shell_trap.c [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+DOUBLE_BATTLE_TEST("Shell Trap activates immediately after being hit on turn 1 and attacks both opponents")
+{
+    GIVEN {
+        ASSUME(gBattleMoves[MOVE_SHELL_TRAP].target == MOVE_TARGET_BOTH);
+        PLAYER(SPECIES_WOBBUFFET) {Speed(1); }
+        PLAYER(SPECIES_WOBBUFFET) {Speed(2); };
+        OPPONENT(SPECIES_WOBBUFFET) {Speed(5); };
+        OPPONENT(SPECIES_WYNAUT) {Speed(1); };
```
Comment (mrgriffin) [inline test/move_effect_shell_trap.c:105]:
```suggestion
        PLAYER(SPECIES_WOBBUFFET) { Speed(1); }
        PLAYER(SPECIES_WOBBUFFET) { Speed(2); }
        OPPONENT(SPECIES_WOBBUFFET) { Speed(5); }
        OPPONENT(SPECIES_WYNAUT) { Speed(1); }
```

File: test/move_effect_shell_trap.c [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    } WHEN {
+        TURN {MOVE(playerLeft, MOVE_SHELL_TRAP);  MOVE(opponentLeft, MOVE_TACKLE, target:playerLeft); MOVE(playerRight, MOVE_CELEBRATE); MOVE(opponentRight, MOVE_CELEBRATE); }
```
Comment (mrgriffin) [inline test/move_effect_shell_trap.c:107]:
```suggestion
        TURN { MOVE(playerLeft, MOVE_SHELL_TRAP); MOVE(opponentLeft, MOVE_TACKLE, target:playerLeft); MOVE(playerRight, MOVE_CELEBRATE); MOVE(opponentRight, MOVE_CELEBRATE); }
```

File: test/move_effect_shell_trap.c [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+DOUBLE_BATTLE_TEST("Shell Trap activates immediately after being hit on turn 2 and attacks both opponents")
+{
+    GIVEN {
+        ASSUME(gBattleMoves[MOVE_SHELL_TRAP].target == MOVE_TARGET_BOTH);
+        PLAYER(SPECIES_WOBBUFFET) {Speed(1); }
+        PLAYER(SPECIES_WOBBUFFET) {Speed(2); };
+        OPPONENT(SPECIES_WOBBUFFET) {Speed(5); };
+        OPPONENT(SPECIES_WYNAUT) {Speed(6); };
```
Comment (mrgriffin) [inline test/move_effect_shell_trap.c:129]:
```suggestion
        PLAYER(SPECIES_WOBBUFFET) { Speed(1); }
        PLAYER(SPECIES_WOBBUFFET) { Speed(2); }
        OPPONENT(SPECIES_WOBBUFFET) { Speed(5); }
        OPPONENT(SPECIES_WYNAUT) { Speed(6); }
```

File: test/move_effect_shell_trap.c [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    } WHEN {
+        TURN {MOVE(playerLeft, MOVE_SHELL_TRAP);  MOVE(opponentLeft, MOVE_TACKLE, target:playerLeft); MOVE(playerRight, MOVE_CELEBRATE); MOVE(opponentRight, MOVE_CELEBRATE); }
```
Comment (mrgriffin) [inline test/move_effect_shell_trap.c:131]:
```suggestion
        TURN { MOVE(playerLeft, MOVE_SHELL_TRAP); MOVE(opponentLeft, MOVE_TACKLE, target:playerLeft); MOVE(playerRight, MOVE_CELEBRATE); MOVE(opponentRight, MOVE_CELEBRATE); }
```

File: test/move_effect_shell_trap.c [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+DOUBLE_BATTLE_TEST("Shell Trap activates immediately after being hit on turn 3 and attacks both opponents")
+{
+    GIVEN {
+        ASSUME(gBattleMoves[MOVE_SHELL_TRAP].target == MOVE_TARGET_BOTH);
+        PLAYER(SPECIES_WOBBUFFET) {Speed(1); }
+        PLAYER(SPECIES_WOBBUFFET) {Speed(7); };
+        OPPONENT(SPECIES_WOBBUFFET) {Speed(5); };
+        OPPONENT(SPECIES_WYNAUT) {Speed(6); };
+    } WHEN {
+        TURN {MOVE(playerLeft, MOVE_SHELL_TRAP);  MOVE(opponentLeft, MOVE_TACKLE, target:playerLeft); MOVE(playerRight, MOVE_CELEBRATE); MOVE(opponentRight, MOVE_CELEBRATE); }
```
Comment (mrgriffin) [inline test/move_effect_shell_trap.c:155]:
```suggestion
        PLAYER(SPECIES_WOBBUFFET) { Speed(1); }
        PLAYER(SPECIES_WOBBUFFET) { Speed(7); }
        OPPONENT(SPECIES_WOBBUFFET) { Speed(5); }
        OPPONENT(SPECIES_WYNAUT) { Speed(6); }
    } WHEN {
        TURN { MOVE(playerLeft, MOVE_SHELL_TRAP); MOVE(opponentLeft, MOVE_TACKLE, target:playerLeft); MOVE(playerRight, MOVE_CELEBRATE); MOVE(opponentRight, MOVE_CELEBRATE); }
```

**Review by mrgriffin — APPROVED** [review]

**Pull request merged**